variables.go

```diff
@@ -137,15 +143,27 @@ func (r *reference) resolveEnv(cfg *Config, opts *options) (string, error) {
 
 func (r *reference) resolve(cfg *Config, opts *options) (value, error) {
 	v, err := r.resolveRef(cfg, opts)
-	if v != nil || err != nil {
+	if v != nil || stopResolving(err) {
 		return v, err
 	}
 
+	previousErr := err
+
 	s, err := r.resolveEnv(cfg, opts)
-	if s == "" || err != nil {
+	if err != nil {
+		// TODO(ph): Not everything is an Error, will do some cleanup in another PR.
+		if v, ok := previousErr.(Error); ok {
+			if v.Reason() == ErrCyclicReference {
+				return nil, previousErr
+			}
+		}
```

Comment:
Nice technique on error predicates:

```
func isCyclicError(err error) bool {
  switch v := err.(type) {
    case Error:
      return v.Reason() == ErrCyclicReference
  }

  // in case we would have special error types with predicates. We don't, so no need to implement this
  type errCyclicPred interface {
    IsCycle() bool
  }
  if v, ok := err.(errCyclicPred); ok {
    return v.IsCycle()
  }

  return false
}

func isMissingError(err error) bool {
  ...
}

func criticalResolveError(err error) bool {
  return !(isCyclicError(err) || isMissingError(err))
}

```

In case we add more error related predicates, common error predicates (like `stopResolving`) can be collected in `errors.go` or `errpred.go`.


In general, we should try to pass `Error` instead of `error` in go-ucfg. Then we would not need to do the type casts...

Reply:
> In general, we should try to pass Error instead of error in go-ucfg. Then we would not need to do the type casts...

Agree concerning the type, but let's do this refactor in other steps :)